server/src/main/java/org/elasticsearch/index/store/StoreFileMetadata.java

```diff
@@ -119,6 +141,24 @@ public boolean isSame(StoreFileMetadata other) {
             // we can't tell if either or is null so we return false in this case! this is why we don't use equals for this!
             return false;
         }
+
+        // If we have the file contents, we directly compare the contents. This is useful to compare segment info
+        // files of source-only snapshots where the original segment info file shares the same id as the source-only
+        // segment info file but its contents are different.
+        if (hashEqualsContents()) {
```

Comment:
Should we compute `hashEqualsContents` once is the constructor and stores it as a class member? It looks like  every time a `StoreFileMetadata` is instanciated we use it.